### ISSUE TITLE
Prevent OOM case (heap memory) by notifying user early

### ISSUE THREAD

**Issue by slrslr** [zlatinb/muwire#59]
Hello,
in other ticket https://github.com/zlatinb/muwire/issues/57 i had an issue with high CPU utilization and at same time low disk activity/not growing number of shared files and app was not responding, when you search for 256 in that thread, it was suggested that built-in memory limit may be the cause.

Some comments:

original one:
> heap is limited to 256mb RAM, which should be good for about ~100,000 shared files

latest one (Jun 4, 2021):
> I was able to do 195k with 1GB

So it may suggest that app should be made either with **higher limit** and the way that it will prevent freezing due to low built-in memory limit and instead:

A) **Notify user before reaching the limit** and possibly stopping operation instructing how to increase
B) limit number of shared files or other memory eating operation and show the memory raising instruction when the limit is near reaching

I was thinking that this behavior would be better than when the app stops responding, because it can make some people to go away thinking app is faulty.

**Comment by zlatinb** [comment]
Hi, thanks for the suggestion.  Unfortunately, there are two problems that make it difficult to implement:

1. The memory status that I can see from inside the application changes very quickly.  Because of the way garbage collection in the JVM works, utilization can be at 99% for one second then back down to 50% the next second.  That makes it difficult to detect accurately when it's really running out of memory.
2. Most MuWire users download the binary bundles from https://muwire.com .  (.exe for Windows, .dmg for Mac, .AppImage for Linux).  Those bundles are built in a different way that makes it impossible to change the memory limit, at all.

Instead, what I can easily do, and might be the right thing, is to remove the memory limit completely.  The positive effect of this is that for as long as there is physical RAM on the machine MuWire will be able to share more files.  The negative effect of this is that, well, there is no limit to how much RAM MuWire could use.

I'm going to test how this behaves when sharing 200k files, and you can help me test too.  Change the `Xmx256M` to `Xms256M` (you're changing the small "x" to "s").  Then just restart MuWire and see how much memory it ends up using during and after it loads the shared files.

**Comment by zlatinb** [comment]
Hi again, with the latest commit I was able to share my 195k files without changing any settings.  MuWire ended up using 1.61GB as reported by the operating system.

**Comment by slrslr** [comment]
Hi,
3 hours back i have built it from git
searched for the 256 in binary file and it was Xms256M
started
it loaded my profile and started loading 350k shared files (in P2P filesharing this is not extraordinary amount when one sharing casual files). By loading i do not mean hashing, it was already done under earlier build.
I have seen
maximum 1.95GB used while under 30K files loaded
after i was around 50K indexed files, i launched "visualvm" and selected muwire process in it
maximum 2.50GB used while around 100K files loaded
maximum 3.35GB used while around 300K files loaded
on previous build while memory was limited to 1.5G(custom limit editing Linux sh binary), app was able to work too.
then there was a big spikes in memory usage seen on the image linked below
after another tens of minutes, i am at 350k files and rough mem. usage (MuWire/Status/System Status) is 12GB
Image from visualVM is [here](https://imgur.com/WTXO2iD.png) (notice increased CPU, GC usage).
It took 3 hours to load all files, maybe during this time no one can download from me. Seems too much work to do on each start of the app (some people start the app maybe daily or more often). Aren't other P2P apps more efficient in this?

**Comment by zlatinb** [comment]
Thanks for trying it out.  The large jump in heap usage as seen in VisualVM is very strange.  I need to understand what caused it.  Two things you can try:
a) Click on "Perform GC"- that should clear some memory and bring down the heap usage.  If it doesn't, something isn't right.
b) Click on "Heap Dump".  On the left, under "Launcher" a new item will appear, called "[heapdump]...".  Right-click on it and select "Save as...".  Then save it as a file somewhere.  You can encrypt that file with my GPG key and give it to me for analysis:
```
gpg --encrypt -r zlatinb@gmail.com < heapdump.hprof > heapdump.hprof.gpg
```

Regarding loading of shared files - that is deliberately slowed down to 500/s because otherwise the UI becomes non-responsive.  However, people should be able to download from you during that time.  If you want to try what happens if files are just loaded as fast as possible, do the following:

1. shut down MuWire
2. open `~/.MuWire/MuWire.properties` in an editor
3. Change `plugin=false` to `plugin=true`
4. Restart MuWire

I expect this will use a lot of memory and CPU, and the UI will be slow to respond until all files finish loading.

Last thing to try regarding memory usage is latest Java available on Manjaro - should be 15 or 16.  Supposedly since Java 13 they made the garbage collector more efficient, but I don't know for sure.

**Comment by zlatinb** [comment]
Two more things to look into:
1. Make sure you have the latest commit from git - `5b3416f`
2. Try running MuWire without VisualVM attached, because sometimes that uses up memory by itself.  (Actually, I don't think Java 15/16 work with VisualVM anyways)

**Comment by slrslr** [comment]
I checked the VisualVM after a few hours and i see the memory usage was down from 12GB to 2GB. After "Perform GC" it is 670MB.

> latest Java available on Manjaro - should be 15 or 16. Supposedly since Java 13 they made the garbage collector more efficient

I have [this](https://bin.disroot.org/?abc7dc2cd7c318fa#DKQZnZEsPqMjhpZhghejjpJGne9BqcjiwiDtSyxx67m4), in short:

> jre-openjdk      [Installed] 15.0.2.u7-1                              extra 
> jre8-openjdk-headless [Installed] 8.u292-1                                 extra 
> jre8-openjdk     [Installed] 8.u292-1        

i have tried to build muwire with recent java package of my distribution, but that [failed, so built with jdk11-openjdk as suggested](https://github.com/zlatinb/muwire/issues/55#issuecomment-852025908).

> Make sure you have the latest commit from git - 5b3416f

The mem. usage shown in previous comment is app version right before you added yours mentioned commit "perform a periodic gc() just in case"

Memory heap dump was sent to gmail, let me know which things would still be handy to try on my end. Thank you

**Comment by zlatinb** [comment]
Ok, I think we're getting closer to an acceptable memory usage.  I've just committed a few more changes that will make the initial loading of files faster, if you set the `plugin=false` property in the `MuWire.properties` file.

I suggest you try the latest code and let MuWire run for a few hours **without** VisualVM attached, and just keep an eye on how much RAM it is using as reported by the operating system.

**Comment by slrslr** [comment]
$ grep plugin ~/.MuWire/MuWire.properties
plugin=true

grep 256 ~/apps/muwire/abc/MuWire-0.8.7/bin/MuWire
DEFAULT_JVM_OPTS='"-Djava.util.logging.config.file=logging.properties" "**-Xms256M**"

started MW, **loading of 350k files was indeed significantly faster**, CPU activity was higher, but maxed out only rarely (good, no other apps lag)

Highest memory usage i have seen 2.9GB
Usual values was like:
Used: 1GB
Total: 3GB
Max: 15GB
Oracle Java 11.0.11
After complete, Used memory dropped to ~750MB. For me these figures on this high RAM computer are not problem, and seems better than limiting 256M heap. [Some details about RAM usage after loading is complete](https://bin.disroot.org/?ed76f6fd6004930c#CkS9KyDQGCsxt33hsjzeDGPSvgVtTd9sasPJcA36dhGe).

CPU during loading:
Load average 20.0-30.0 (CPU 50%-100% utilized, not just by MW)
CPU graph showing like 100% CPU usage increase for the 35 minutes when MuWire was loading files. [HERE graph](https://i.postimg.cc/8PCM3NvZ/cpu-muwire.png)
200%-700% java processes CPU usage (shown by htop) during that 35 minutes startup/loading files :-/

Some other non-java P2P apps seems to be faster/less resource hungry working/loading shared files upon start.

If you want i try something, let me know, thank you.

**Comment by zlatinb** [comment]
Hey, thanks for the good news!  I think the memory usage is under control now and we can close this issue.

The loading of shared files on startup is a tricky one; if `plugin=true` the UI starts to lag by a few seconds.  So I think I will make it a configuration option for advanced users, with default being throttled.

The system load average is high because MuWire tries to load many files from disk in parallel.  Maybe that's not really helping.  If you want to keep testing further, please open a new issue for speeding up the load time and we'll do some experiments there.

**Comment by slrslr** [comment]
Does it mean that 256 heap memory limit will no longer be used or should no longer cause any issue to new users in default app configuration and hundreds of thousands of shared files? Please close this if it should be reasonably fine.

**Comment by zlatinb** [comment]
The 256 will no longer be used.  By default there will be no heap limit, but there are going to be periodic garbage collections (once every minute) to make sure MuWire does not use up too much memory.  Users should be able to share hundreds of thousands of files with default settings.

**Comment by slrslr** [comment]
> The 256 will no longer be used

Hello @zlatinb i have been using Xmx1536M in muwire for months and i was happy with the UI speed and loading of 500K shared files (in Library) was not too slow. Overall seen no issue. Now after i have built from Arch Linux AUR muwire-git, it is using Xmx256M and loading of shared files is super slow and interface is also significantly slower during the process.
So i had to quit and run:
sudo sed -i "s/Xmx256M/Xmx1536M/g" /usr/bin/muwire
(and do this manually every time muwire-git AUR is built by updater)

So letting you know to consider if can be good to raise default value.
